You will have to import five different datasets. Each one is an exchange traded fund (etf) offered by Vanguard that hold a collection of securities (stocks). You could technically buy one of these and own all the businesses that are part of that etf. Pretty cool.

The composition of each etf is market capitalization weighted. That means that the bigger the company is in terms of market value (think Apple with its trillion dollar market cap), the more of your dollars that you invest in that etf goes to this one stock. That is shown below.

![Large companies dominate...](large-companies-dominate-total-stock-market-index.png)

But you don't want that. What you want is this...

![A more proportional portfolio...](spreading-your-dollars-across-all-size-companies.png)

Let's explore one of those etfs.

In [1]:
import numpy as np

In [2]:
import pandas as pd
pd.read_csv('vti_12_31_2019.csv').head(10)

Unnamed: 0        date                  holding   marketValue
0           0  2019-12-31          Microsoft Corp.  3.388885e+10
1           1  2019-12-31               Apple Inc.  3.307893e+10
2           2  2019-12-31          Amazon.com Inc.  2.193640e+10
3           3  2019-12-31            Facebook Inc.  1.391406e+10
4           4  2019-12-31  Berkshire Hathaway Inc.  1.208788e+10
5           5  2019-12-31     JPMorgan Chase & Co.  1.157386e+10
6           6  2019-12-31            Alphabet Inc.  1.129601e+10
7           7  2019-12-31    Alphabet Inc. Class C  1.100778e+10
8           8  2019-12-31        Johnson & Johnson  1.081596e+10
9           9  2019-12-31                Visa Inc.  9.063540e+09

So these are the top ten holdings within this etf. The marketValue column for each holding tells us the amount of money that goes towards that holding if we were to make an investment in this etf. To find % ownership for each holding, you can divide the marketValue for that holding with the total sum of marketValues for all holdings in that etf.

A brief discreption on the etfs...

[VTI](https://investor.vanguard.com/etf/profile/VTI) - Total stock market etf. It technically owns all the publicly traded companies in these United States.

[VO](https://investor.vanguard.com/etf/profile/vo) - Mid-size company etf. It owns companies that are not too larger or small.

[VB](https://investor.vanguard.com/etf/profile/vb) - Small-size company etf. It owns companies that are small in market value.

[VOO](https://investor.vanguard.com/etf/profile/voo) - These are all the companies that are part of the S&P 500 index. This is basically a large cap index and almost all companies within this are also found in VTI.

[VXF](https://investor.vanguard.com/etf/profile/vxf) - These are all the companies that are part of VTI but excluding those that are found in the S&P 500 index. Which means VXF + VOO = VTI.

### Question

Say you had a 100 dollars to invest and you were to split that equally between these 5 etfs. How will your money be divvied up between the companies within that combined portfolio (hint: use merge in combining datasets). 

- What portion of your money will go to the top 50 companies?
- What portion to the next 450 companies?
- And what to the remaining set of companies?

Visualizations come in handy but are not necessary. Be as creative as possible. There might be overlap in holdings between etfs and you might have to use regular expressions to clean some things out to match holdings.

## <b>ANSWER: To find how the money 100 dollars is divided in investing between companies within that combined portfolio, there will create a new column 'weight' which represents the weight of each company in each of 5 etfs. Next, there will create another column 'comb_weight' to represent the weight of that company in the combined portfolio. Since 100 dollars is equally split between 5 etfs, the 'comb_weight' column can be computed by divding the individual etf weight by 5. </b>

In [3]:
dfvb = pd.read_csv('vb_12_31_2019.csv')

In [4]:
dfvb['weight'] = dfvb.marketValue / dfvb.marketValue.sum()

In [5]:
dfvb['comb_weight'] = dfvb.weight / 5

In [6]:
dfvb.drop(['Unnamed: 0', 'date'], axis=1, inplace=True)


## <b>Adding columns of calculated weight and comb_weight for VB.</b> 

In [7]:
dfvb

holding  marketValue        weight  \
0                  Leidos Holdings Inc.  347521932.0  3.625613e-03   
1              Zebra Technologies Corp.  344245760.0  3.591434e-03   
2                    Atmos Energy Corp.  334201885.0  3.486648e-03   
3                            IDEX Corp.  327868292.0  3.420571e-03   
4                            STERIS plc  322979504.0  3.369568e-03   
...                                 ...          ...           ...   
1352          AMAG Pharmaceuticals Inc.      33139.0  3.457313e-07   
1353  Altisource Portfolio Solutions SA      19775.0  2.063078e-07   
1354       Synergy Pharmaceuticals Inc.      12807.0  1.336124e-07   
1355                       Ashford Inc.        614.0  6.405715e-09   
1356                    Dorian LPG Ltd.        310.0  3.234156e-09   

       comb_weight  
0     7.251226e-04  
1     7.182867e-04  
2     6.973297e-04  
3     6.841143e-04  
4     6.739136e-04  
...            ...  
1352  6.914625e-08  
1353  4.126157e-08  
1354  2.672247e-08  
1355  1.281143e-09  
1356  6.468312e-10  

[1357 rows x 4 columns]

In [8]:
dfvo = pd.read_csv('vo_12_31_2019.csv')

In [9]:
dfvo['weight'] = dfvo.marketValue / dfvo.marketValue.sum()

In [10]:
dfvo['comb_weight'] = dfvo.weight / 5

In [11]:
dfvo.drop(['Unnamed: 0', 'date'], axis=1, inplace=True)


## <b>Adding columns of calculated weight and comb_weight for VO.</b> 

In [12]:
dfvo

holding  marketValue        weight   comb_weight
0         Newmont Goldcorp Corp.  896765462.0  8.060491e-03  1.612098e-03
1         Amphenol Corp. Class A  807963791.0  7.262306e-03  1.452461e-03
2                     ONEOK Inc.  787016807.0  7.074026e-03  1.414805e-03
3          WEC Energy Group Inc.  732058286.0  6.580036e-03  1.316007e-03
4                IHS Markit Ltd.  722329137.0  6.492587e-03  1.298517e-03
..                           ...          ...           ...           ...
334   Levi Strauss & Co. Class A   21514831.0  1.933840e-04  3.867680e-05
335         Lennar Corp. Class B    9520251.0  8.557187e-05  1.711437e-05
336       ViacomCBS Inc. Class A    6668086.0  5.993546e-05  1.198709e-05
337           News Corp. Class B    5167272.0  4.644553e-05  9.289107e-06
338  Alliance Data Systems Corp.      23338.0  2.097714e-07  4.195428e-08

[339 rows x 4 columns]

In [13]:
dfvoo = pd.read_csv('voo_12_31_2019.csv')

In [14]:
dfvoo['weight'] = dfvoo.marketValue / dfvoo.marketValue.sum()

In [15]:
dfvoo['comb_weight'] = dfvoo.weight / 5

In [16]:
dfvoo.drop(['Unnamed: 0', 'date'], axis=1, inplace=True)


## <b>Adding columns of calculated weight and comb_weight for VOO.</b> 

In [17]:
dfvoo

holding   marketValue        weight   comb_weight
0                       Apple Inc.  2.450349e+10  4.585529e-02  9.171059e-03
1                  Microsoft Corp.  2.403557e+10  4.497965e-02  8.995931e-03
2                  Amazon.com Inc.  1.537491e+10  2.877227e-02  5.754454e-03
3            Facebook Inc. Class A  9.868079e+09  1.846691e-02  3.693382e-03
4             JPMorgan Chase & Co.  8.651537e+09  1.619030e-02  3.238060e-03
..                             ...           ...           ...           ...
504              Coty Inc. Class A  6.657366e+07  1.245845e-04  2.491690e-05
505                   Macerich Co.  3.955086e+06  7.401463e-06  1.480293e-06
506           Lennar Corp. Class B  3.440783e+06  6.439007e-06  1.287801e-06
507             News Corp. Class B  2.135730e+05  3.996759e-07  7.993518e-08
508  Wyndham Hotels & Resorts Inc.  1.356700e+04  2.538899e-08  5.077798e-09

[509 rows x 4 columns]

In [18]:
dfvoo.marketValue.sum()

534365463237.0

In [19]:
dfvti = pd.read_csv('vti_12_31_2019.csv')

In [20]:
dfvti['weight'] = dfvti.marketValue / dfvti.marketValue.sum()

In [21]:
dfvti['comb_weight'] = dfvti.weight / 5

In [22]:
dfvti.drop(['Unnamed: 0', 'date'], axis=1, inplace=True)


## <b>Adding columns of calculated weight and comb_weight for VTI.</b> 

In [23]:
dfvti

holding   marketValue        weight   comb_weight
0                    Microsoft Corp.  3.388885e+10  3.786959e-02  7.573918e-03
1                         Apple Inc.  3.307893e+10  3.696454e-02  7.392908e-03
2                    Amazon.com Inc.  2.193640e+10  2.451315e-02  4.902631e-03
3                      Facebook Inc.  1.391406e+10  1.554847e-02  3.109694e-03
4            Berkshire Hathaway Inc.  1.208788e+10  1.350778e-02  2.701556e-03
...                              ...           ...           ...           ...
3545       Jones Energy Inc. Class A  8.000000e+00  8.939718e-12  1.787944e-12
3546             THESTREET Inc. CVR1  2.000000e+00  2.234929e-12  4.469859e-13
3547  Harvest Natural Resources Inc.  1.000000e+00  1.117465e-12  2.234929e-13
3548          Universal Travel Group  0.000000e+00  0.000000e+00  0.000000e+00
3549                Rex Energy Corp.  0.000000e+00  0.000000e+00  0.000000e+00

[3550 rows x 4 columns]

In [24]:
dfvti.marketValue.sum()

894882840379.0

In [25]:
dfvti.marketValue.sum() + dfvoo.marketValue.sum()

1429248303616.0

In [26]:
dfvxf = pd.read_csv('vxf_12_31_2019.csv')

In [27]:
dfvxf['weight'] = dfvxf.marketValue/dfvxf.marketValue.sum()

In [28]:
dfvxf['comb_weight'] = dfvxf.weight / 5

In [29]:
dfvxf.drop(['Unnamed: 0', 'date'], axis=1, inplace=True)


## <b>Adding columns of calculated weight and comb_weight for VXF.</b> 

In [30]:
dfvxf

holding  marketValue        weight  \
0                            Tesla Inc.  844923272.0  1.145656e-02   
1                   Blackstone Group LP  524426088.0  7.110846e-03   
2              Lululemon Athletica Inc.  393325388.0  5.333214e-03   
3                  Workday Inc. Class A  359631265.0  4.876345e-03   
4               Palo Alto Networks Inc.  321314013.0  4.356791e-03   
...                                 ...          ...           ...   
3244  Priority Technology Holdings Inc.        260.0  3.525416e-09   
3245                   Cohen & Co. Inc.        253.0  3.430501e-09   
3246        InnSuites Hospitality Trust        165.0  2.237283e-09   
3247     Harvest Natural Resources Inc.          1.0  1.355929e-11   
3248                      TheStreet CVR          1.0  1.355929e-11   

       comb_weight  
0     2.291312e-03  
1     1.422169e-03  
2     1.066643e-03  
3     9.752691e-04  
4     8.713581e-04  
...            ...  
3244  7.050832e-10  
3245  6.861002e-10  
3246  4.474566e-10  
3247  2.711858e-12  
3248  2.711858e-12  

[3249 rows x 4 columns]

## <b>For a test, merge VOO (509 companies) and VTI (3550 companies) dataset.</b> 

In [31]:
dftest = dfvoo.merge(dfvti, how='outer')

In [32]:
dftest

holding   marketValue        weight   comb_weight
0                         Apple Inc.  2.450349e+10  4.585529e-02  9.171059e-03
1                    Microsoft Corp.  2.403557e+10  4.497965e-02  8.995931e-03
2                    Amazon.com Inc.  1.537491e+10  2.877227e-02  5.754454e-03
3              Facebook Inc. Class A  9.868079e+09  1.846691e-02  3.693382e-03
4               JPMorgan Chase & Co.  8.651537e+09  1.619030e-02  3.238060e-03
...                              ...           ...           ...           ...
4054       Jones Energy Inc. Class A  8.000000e+00  8.939718e-12  1.787944e-12
4055             THESTREET Inc. CVR1  2.000000e+00  2.234929e-12  4.469859e-13
4056  Harvest Natural Resources Inc.  1.000000e+00  1.117465e-12  2.234929e-13
4057          Universal Travel Group  0.000000e+00  0.000000e+00  0.000000e+00
4058                Rex Energy Corp.  0.000000e+00  0.000000e+00  0.000000e+00

[4059 rows x 4 columns]

In [33]:
dftest['comb_weight'].sum()

0.39999999999999997

In [34]:
dftest.loc[ dftest.loc[:, 'holding']=='Apple Inc.', :]

holding   marketValue    weight  comb_weight
0    Apple Inc.  2.450349e+10  0.045855     0.009171
510  Apple Inc.  3.307893e+10  0.036965     0.007393

In [35]:
dftest.loc[ dftest.loc[:, 'holding']=='Apple Inc.', 'comb_weight'].sum()

0.01656396737833214

## <b>From the test, the entries from two dataset were combined but for the same company, there are still separated entries, i.e. the total weight of that company is not represented until you add up the individual weight.</b> 

## <b>There needs to merge two or more entries of the same company and add up the individual weight to become combined overall weight of that company in the combined portfolio.</b> 

## <b>First, drop the column marketValue and column weight as we only add up the column comb_weight only after dataset merge.</b> 

In [36]:
dfvb2 = dfvb.drop(['marketValue', 'weight'], axis=1)

In [37]:
dfvb2


holding   comb_weight
0                  Leidos Holdings Inc.  7.251226e-04
1              Zebra Technologies Corp.  7.182867e-04
2                    Atmos Energy Corp.  6.973297e-04
3                            IDEX Corp.  6.841143e-04
4                            STERIS plc  6.739136e-04
...                                 ...           ...
1352          AMAG Pharmaceuticals Inc.  6.914625e-08
1353  Altisource Portfolio Solutions SA  4.126157e-08
1354       Synergy Pharmaceuticals Inc.  2.672247e-08
1355                       Ashford Inc.  1.281143e-09
1356                    Dorian LPG Ltd.  6.468312e-10

[1357 rows x 2 columns]

In [38]:
dfvo2 = dfvo.drop(['marketValue', 'weight'], axis=1)


In [39]:
dfvo2

holding   comb_weight
0         Newmont Goldcorp Corp.  1.612098e-03
1         Amphenol Corp. Class A  1.452461e-03
2                     ONEOK Inc.  1.414805e-03
3          WEC Energy Group Inc.  1.316007e-03
4                IHS Markit Ltd.  1.298517e-03
..                           ...           ...
334   Levi Strauss & Co. Class A  3.867680e-05
335         Lennar Corp. Class B  1.711437e-05
336       ViacomCBS Inc. Class A  1.198709e-05
337           News Corp. Class B  9.289107e-06
338  Alliance Data Systems Corp.  4.195428e-08

[339 rows x 2 columns]

In [40]:
dfvoo2 = dfvoo.drop(['marketValue', 'weight'], axis=1)


In [41]:
dfvoo2

holding   comb_weight
0                       Apple Inc.  9.171059e-03
1                  Microsoft Corp.  8.995931e-03
2                  Amazon.com Inc.  5.754454e-03
3            Facebook Inc. Class A  3.693382e-03
4             JPMorgan Chase & Co.  3.238060e-03
..                             ...           ...
504              Coty Inc. Class A  2.491690e-05
505                   Macerich Co.  1.480293e-06
506           Lennar Corp. Class B  1.287801e-06
507             News Corp. Class B  7.993518e-08
508  Wyndham Hotels & Resorts Inc.  5.077798e-09

[509 rows x 2 columns]

In [42]:
dfvti2 = dfvti.drop(['marketValue', 'weight'], axis=1)


In [43]:
dfvti2

holding   comb_weight
0                    Microsoft Corp.  7.573918e-03
1                         Apple Inc.  7.392908e-03
2                    Amazon.com Inc.  4.902631e-03
3                      Facebook Inc.  3.109694e-03
4            Berkshire Hathaway Inc.  2.701556e-03
...                              ...           ...
3545       Jones Energy Inc. Class A  1.787944e-12
3546             THESTREET Inc. CVR1  4.469859e-13
3547  Harvest Natural Resources Inc.  2.234929e-13
3548          Universal Travel Group  0.000000e+00
3549                Rex Energy Corp.  0.000000e+00

[3550 rows x 2 columns]

In [44]:
dfvxf2 = dfvxf.drop(['marketValue', 'weight'], axis=1)


In [45]:
dfvxf2

holding   comb_weight
0                            Tesla Inc.  2.291312e-03
1                   Blackstone Group LP  1.422169e-03
2              Lululemon Athletica Inc.  1.066643e-03
3                  Workday Inc. Class A  9.752691e-04
4               Palo Alto Networks Inc.  8.713581e-04
...                                 ...           ...
3244  Priority Technology Holdings Inc.  7.050832e-10
3245                   Cohen & Co. Inc.  6.861002e-10
3246        InnSuites Hospitality Trust  4.474566e-10
3247     Harvest Natural Resources Inc.  2.711858e-12
3248                      TheStreet CVR  2.711858e-12

[3249 rows x 2 columns]

## <b>Then, merging VOO and VTI...

In [46]:
df_voo_vti = dfvoo2.merge(dfvti2, how='outer', left_on='holding', right_on='holding' )

In [47]:
df_voo_vti

holding  comb_weight_x  comb_weight_y
0                         Apple Inc.       0.009171   7.392908e-03
1                    Microsoft Corp.       0.008996   7.573918e-03
2                    Amazon.com Inc.       0.005754   4.902631e-03
3              Facebook Inc. Class A       0.003693            NaN
4               JPMorgan Chase & Co.       0.003238   2.586676e-03
...                              ...            ...            ...
3583       Jones Energy Inc. Class A            NaN   1.787944e-12
3584             THESTREET Inc. CVR1            NaN   4.469859e-13
3585  Harvest Natural Resources Inc.            NaN   2.234929e-13
3586          Universal Travel Group            NaN   0.000000e+00
3587                Rex Energy Corp.            NaN   0.000000e+00

[3588 rows x 3 columns]

In [48]:
df_voo_vti = df_voo_vti.groupby(np.where(df_voo_vti.columns == 'holding', 'holding', 'comb_weight'), axis=1).sum() 

In [49]:
df_voo_vti

comb_weight                         holding
0     1.656397e-02                      Apple Inc.
1     1.656985e-02                 Microsoft Corp.
2     1.065708e-02                 Amazon.com Inc.
3     3.693382e-03           Facebook Inc. Class A
4     5.824736e-03            JPMorgan Chase & Co.
...            ...                             ...
3583  1.787944e-12       Jones Energy Inc. Class A
3584  4.469859e-13             THESTREET Inc. CVR1
3585  2.234929e-13  Harvest Natural Resources Inc.
3586  0.000000e+00          Universal Travel Group
3587  0.000000e+00                Rex Energy Corp.

[3588 rows x 2 columns]

## <b>Verifying the combined weight of 2 ETFs should be added up to 0.4. 

In [50]:
df_voo_vti['comb_weight'].sum()

0.40007643360211304

## <b>Next, merging VXF with VOO and VTI...

In [51]:
df_vxf_voo_vti = df_voo_vti.merge(dfvxf2, how='outer', left_on='holding', right_on='holding' )

In [52]:
df_vxf_voo_vti

comb_weight_x                            holding  comb_weight_y
0          0.016564                         Apple Inc.            NaN
1          0.016570                    Microsoft Corp.            NaN
2          0.010657                    Amazon.com Inc.            NaN
3          0.003693              Facebook Inc. Class A            NaN
4          0.005825               JPMorgan Chase & Co.            NaN
...             ...                                ...            ...
4073            NaN                       Nephros Inc.   2.698299e-09
4074            NaN  Priority Technology Holdings Inc.   7.050832e-10
4075            NaN                   Cohen & Co. Inc.   6.861002e-10
4076            NaN        InnSuites Hospitality Trust   4.474566e-10
4077            NaN                      TheStreet CVR   2.711858e-12

[4078 rows x 3 columns]

In [53]:
df_vxf_voo_vti = df_vxf_voo_vti.groupby(np.where(df_vxf_voo_vti.columns == 'holding', 'holding', 'comb_weight'), axis=1).sum() 

In [54]:
df_vxf_voo_vti

comb_weight                            holding
0     1.656397e-02                         Apple Inc.
1     1.656985e-02                    Microsoft Corp.
2     1.065708e-02                    Amazon.com Inc.
3     3.693382e-03              Facebook Inc. Class A
4     5.824736e-03               JPMorgan Chase & Co.
...            ...                                ...
4073  2.698299e-09                       Nephros Inc.
4074  7.050832e-10  Priority Technology Holdings Inc.
4075  6.861002e-10                   Cohen & Co. Inc.
4076  4.474566e-10        InnSuites Hospitality Trust
4077  2.711858e-12                      TheStreet CVR

[4078 rows x 2 columns]

## <b>Verifying the combined weight of 3 ETFs should be added up to 0.6. 

In [55]:
df_vxf_voo_vti['comb_weight'].sum()

0.6016874153433667

## <b>Merging VB with VOO, VTI and VXF...

In [56]:
df_vb_vxf_voo_vti = df_vxf_voo_vti.merge(dfvb2, how='outer', left_on='holding', right_on='holding' )

In [57]:
df_vb_vxf_voo_vti

comb_weight_x                            holding  comb_weight_y
0      1.656397e-02                         Apple Inc.            NaN
1      1.656985e-02                    Microsoft Corp.            NaN
2      1.065708e-02                    Amazon.com Inc.            NaN
3      3.693382e-03              Facebook Inc. Class A            NaN
4      5.824736e-03               JPMorgan Chase & Co.            NaN
...             ...                                ...            ...
4076   7.050832e-10  Priority Technology Holdings Inc.            NaN
4077   6.861002e-10                   Cohen & Co. Inc.            NaN
4078   4.474566e-10        InnSuites Hospitality Trust            NaN
4079   2.711858e-12                      TheStreet CVR            NaN
4080            NaN                        Intelsat SA       0.000029

[4081 rows x 3 columns]

In [58]:
df_vb_vxf_voo_vti = df_vb_vxf_voo_vti.groupby(np.where(df_vb_vxf_voo_vti.columns == 'holding', 'holding', 'comb_weight'), axis=1).sum() 

In [59]:
df_vb_vxf_voo_vti

comb_weight                            holding
0     1.656397e-02                         Apple Inc.
1     1.656985e-02                    Microsoft Corp.
2     1.065708e-02                    Amazon.com Inc.
3     3.693382e-03              Facebook Inc. Class A
4     5.824736e-03               JPMorgan Chase & Co.
...            ...                                ...
4076  7.050832e-10  Priority Technology Holdings Inc.
4077  6.861002e-10                   Cohen & Co. Inc.
4078  4.474566e-10        InnSuites Hospitality Trust
4079  2.711858e-12                      TheStreet CVR
4080  2.865675e-05                        Intelsat SA

[4081 rows x 2 columns]

## <b>Verifying the combined weight of 4 ETFs should be added up to 0.8. 

In [60]:
df_vb_vxf_voo_vti['comb_weight'].sum()

0.8026409969066777

## <b>Merging VO with VOO, VTI, VXF and VB, hence, all 5 ETFs together...

In [61]:
df_all = df_vb_vxf_voo_vti.merge(dfvo2, how='outer', left_on='holding', right_on='holding' )

In [62]:
df_all

comb_weight_x                            holding  comb_weight_y
0      1.656397e-02                         Apple Inc.            NaN
1      1.656985e-02                    Microsoft Corp.            NaN
2      1.065708e-02                    Amazon.com Inc.            NaN
3      3.693382e-03              Facebook Inc. Class A            NaN
4      5.824736e-03               JPMorgan Chase & Co.            NaN
...             ...                                ...            ...
4076   7.050832e-10  Priority Technology Holdings Inc.            NaN
4077   6.861002e-10                   Cohen & Co. Inc.            NaN
4078   4.474566e-10        InnSuites Hospitality Trust            NaN
4079   2.711858e-12                      TheStreet CVR            NaN
4080   2.865675e-05                        Intelsat SA            NaN

[4081 rows x 3 columns]

In [63]:
df_all = df_all.groupby(np.where(df_all.columns == 'holding', 'holding', 'comb_weight'), axis=1).sum() 

## <b>Now we have the combined merged dataset containing the overall weight of each company in the combined portfolio. </b>

In [64]:
df_all

comb_weight                            holding
0     1.656397e-02                         Apple Inc.
1     1.656985e-02                    Microsoft Corp.
2     1.065708e-02                    Amazon.com Inc.
3     3.693382e-03              Facebook Inc. Class A
4     5.824736e-03               JPMorgan Chase & Co.
...            ...                                ...
4076  7.050832e-10  Priority Technology Holdings Inc.
4077  6.861002e-10                   Cohen & Co. Inc.
4078  4.474566e-10        InnSuites Hospitality Trust
4079  2.711858e-12                      TheStreet CVR
4080  2.865675e-05                        Intelsat SA

[4081 rows x 2 columns]

## <b>Verifying the combined weight of all funds should be added up to 1. 

In [65]:
df_all['comb_weight'].sum()

1.0041349653931042

## <b>Sorting the combined dataset with the combined weight in descending order.</b> 

In [66]:
df_all.sort_values(by=['comb_weight'],ascending=False, inplace=True)

## <b>Renaming index after sort</b>

In [67]:
df_all['Index'] = np.arange(df_all.shape[0])
df_all.set_index('Index', inplace=True)

In [68]:
df_all

comb_weight                    holding
Index                                         
0      1.656985e-02            Microsoft Corp.
1      1.656397e-02                 Apple Inc.
2      1.065708e-02            Amazon.com Inc.
3      5.824736e-03       JPMorgan Chase & Co.
4      5.464071e-03      Alphabet Inc. Class C
...             ...                        ...
4076   2.711858e-12              TheStreet CVR
4077   1.787944e-12  Jones Energy Inc. Class A
4078   4.469859e-13        THESTREET Inc. CVR1
4079   0.000000e+00     Universal Travel Group
4080   0.000000e+00           Rex Energy Corp.

[4081 rows x 2 columns]

## <b>Here we can use the combined dataset to answer those questions.</b>

- What portion of your money will go to the top 50 companies?
- What portion to the next 450 companies?
- And what to the remaining set of companies?

In [69]:
df_all.nlargest(50, 'comb_weight')


comb_weight                          holding
Index                                              
0         0.016570                  Microsoft Corp.
1         0.016564                       Apple Inc.
2         0.010657                  Amazon.com Inc.
3         0.005825             JPMorgan Chase & Co.
4         0.005464            Alphabet Inc. Class C
5         0.005288                Johnson & Johnson
6         0.004290             Procter & Gamble Co.
7         0.004067                Exxon Mobil Corp.
8         0.003932                        AT&T Inc.
9         0.003927            Bank of America Corp.
10        0.003836          UnitedHealth Group Inc.
11        0.003693            Facebook Inc. Class A
12        0.003590                  Walt Disney Co.
13        0.003586                      Intel Corp.
14        0.003497      Verizon Communications Inc.
15        0.003281                  Home Depot Inc.
16        0.003218  Berkshire Hathaway Inc. Class B
17        0.003190                 Merck & Co. Inc.
18        0.003139                    Chevron Corp.
19        0.003110                    Facebook Inc.
20        0.002987                      Pfizer Inc.
21        0.002985            Alphabet Inc. Class A
22        0.002940                    Coca-Cola Co.
23        0.002838                Wells Fargo & Co.
24        0.002802               Cisco Systems Inc.
25        0.002702          Berkshire Hathaway Inc.
26        0.002671                       Tesla Inc.
27        0.002625                     PepsiCo Inc.
28        0.002525                    Alphabet Inc.
29        0.002464         Lululemon Athletica Inc.
30        0.002405                Visa Inc. Class A
31        0.002402                   Citigroup Inc.
32        0.002399                       Boeing Co.
33        0.002325                     Walmart Inc.
34        0.002199                       Adobe Inc.
35        0.002115              Abbott Laboratories
36        0.002103           Newmont Goldcorp Corp.
37        0.002094                    Medtronic plc
38        0.002071         Bristol-Myers Squibb Co.
39        0.002050                 McDonald's Corp.
40        0.002026                        Visa Inc.
41        0.001989                      Splunk Inc.
42        0.001982          Mastercard Inc. Class A
43        0.001980          Palo Alto Networks Inc.
44        0.001973                       Amgen Inc.
45        0.001973                CoStar Group Inc.
46        0.001953                     Netflix Inc.
47        0.001938                     NVIDIA Corp.
48        0.001932              salesforce.com Inc.
49        0.001895           Amphenol Corp. Class A

In [70]:
top50 = df_all.loc[:, 'comb_weight'].nlargest(50).sum()
top50*100

18.00625316482548

## <b>ANSWER: 18% of money will go to the top 50 companies.</b>

In [71]:
top500 = df_all.loc[:, 'comb_weight'].nlargest(500).sum()
top500

0.6170595348893017

In [72]:
top450 = top500 - top50
top450*100

43.69970032410468

## <b>ANSWER: 44% of money will go to the next 450 companies.</b>

In [73]:
all = df_all.loc[:, 'comb_weight'].sum()
all

1.0041349653931042

In [74]:
all = df_all.loc[:, 'comb_weight'].sum()
rest = all - top500
rest

0.38707543050380255

## <b>ANSWER: 38% of money will go to the remaining set of companies.</b>

### <b>Question</b>

Try some other proportions out in how you will divvy up your portfolio to get broader exposure to more mid and small size companies. Provide your best allocation and state the reasons why. This is a more open-ended question so be as creative as possible.

## <b>Mid-size companies</b>

In [75]:
dfvo

holding  marketValue        weight   comb_weight
0         Newmont Goldcorp Corp.  896765462.0  8.060491e-03  1.612098e-03
1         Amphenol Corp. Class A  807963791.0  7.262306e-03  1.452461e-03
2                     ONEOK Inc.  787016807.0  7.074026e-03  1.414805e-03
3          WEC Energy Group Inc.  732058286.0  6.580036e-03  1.316007e-03
4                IHS Markit Ltd.  722329137.0  6.492587e-03  1.298517e-03
..                           ...          ...           ...           ...
334   Levi Strauss & Co. Class A   21514831.0  1.933840e-04  3.867680e-05
335         Lennar Corp. Class B    9520251.0  8.557187e-05  1.711437e-05
336       ViacomCBS Inc. Class A    6668086.0  5.993546e-05  1.198709e-05
337           News Corp. Class B    5167272.0  4.644553e-05  9.289107e-06
338  Alliance Data Systems Corp.      23338.0  2.097714e-07  4.195428e-08

[339 rows x 4 columns]

## <b>Small-size companies</b>

In [76]:
dfvb

holding  marketValue        weight  \
0                  Leidos Holdings Inc.  347521932.0  3.625613e-03   
1              Zebra Technologies Corp.  344245760.0  3.591434e-03   
2                    Atmos Energy Corp.  334201885.0  3.486648e-03   
3                            IDEX Corp.  327868292.0  3.420571e-03   
4                            STERIS plc  322979504.0  3.369568e-03   
...                                 ...          ...           ...   
1352          AMAG Pharmaceuticals Inc.      33139.0  3.457313e-07   
1353  Altisource Portfolio Solutions SA      19775.0  2.063078e-07   
1354       Synergy Pharmaceuticals Inc.      12807.0  1.336124e-07   
1355                       Ashford Inc.        614.0  6.405715e-09   
1356                    Dorian LPG Ltd.        310.0  3.234156e-09   

       comb_weight  
0     7.251226e-04  
1     7.182867e-04  
2     6.973297e-04  
3     6.841143e-04  
4     6.739136e-04  
...            ...  
1352  6.914625e-08  
1353  4.126157e-08  
1354  2.672247e-08  
1355  1.281143e-09  
1356  6.468312e-10  

[1357 rows x 4 columns]

## <b>Proporation of ETF also containing Mid-size companies</b>

In [77]:
df_vo_vb = dfvo.merge(dfvb, how='inner', left_on='holding', right_on='holding' )

In [78]:
df_vo_vb

holding  marketValue_x      weight_x  \
0           Burlington Stores Inc.    190030853.0  1.708074e-03   
1         RingCentral Inc. Class A    158720663.0  1.426646e-03   
2     Jack Henry & Associates Inc.    141039879.0  1.267724e-03   
3                Vail Resorts Inc.    121630024.0  1.093260e-03   
4               DXC Technology Co.    120928383.0  1.086953e-03   
5                  NRG Energy Inc.    119544627.0  1.074516e-03   
6                     Cognex Corp.    114316444.0  1.027522e-03   
7          Jones Lang LaSalle Inc.    112707607.0  1.013062e-03   
8                      Pentair plc     96885880.0  8.708495e-04   
9    Reliance Steel & Aluminum Co.     95332313.0  8.568854e-04   
10  People's United Financial Inc.     94077095.0  8.456030e-04   
11              XPO Logistics Inc.     92421156.0  8.307188e-04   
12           Cabot Oil & Gas Corp.     89336943.0  8.029966e-04   
13               FLIR Systems Inc.     87826573.0  7.894208e-04   
14  Robert Half International Inc.     87477338.0  7.862817e-04   
15                  AO Smith Corp.     82061091.0  7.375983e-04   
16     Axalta Coating Systems Ltd.     80769730.0  7.259910e-04   
17             Carvana Co. Class A     58196403.0  5.230928e-04   
18     Alliance Data Systems Corp.        23338.0  2.097714e-07   

    comb_weight_x  marketValue_y  weight_y  comb_weight_y  
0    3.416149e-04    188613190.0  0.001968       0.000394  
1    2.853291e-04    157512817.0  0.001643       0.000329  
2    2.535447e-04    142529937.0  0.001487       0.000297  
3    2.186520e-04    122721011.0  0.001280       0.000256  
4    2.173907e-04    119153158.0  0.001243       0.000249  
5    2.149031e-04    118817321.0  0.001240       0.000248  
6    2.055045e-04    112210181.0  0.001171       0.000234  
7    2.026123e-04    114267801.0  0.001192       0.000238  
8    1.741699e-04     97582003.0  0.001018       0.000204  
9    1.713771e-04     96732068.0  0.001009       0.000202  
10   1.691206e-04     95936534.0  0.001001       0.000200  
11   1.661438e-04     91110729.0  0.000951       0.000190  
12   1.605993e-04     88123936.0  0.000919       0.000184  
13   1.578842e-04     86586786.0  0.000903       0.000181  
14   1.572563e-04     88143696.0  0.000920       0.000184  
15   1.475197e-04     80515078.0  0.000840       0.000168  
16   1.451982e-04     81578248.0  0.000851       0.000170  
17   1.046186e-04     58232303.0  0.000608       0.000122  
18   4.195428e-08    122283863.0  0.001276       0.000255

In [79]:
df_vo_vb.holding.count() / dfvo.holding.count()

0.05604719764011799

In [80]:
df_vo_voo = dfvo.merge(dfvoo, how='inner', left_on='holding', right_on='holding' )

In [81]:
df_vo_voo

holding  marketValue_x      weight_x  comb_weight_x  \
0         Newmont Goldcorp Corp.    896765462.0  8.060491e-03   1.612098e-03   
1         Amphenol Corp. Class A    807963791.0  7.262306e-03   1.452461e-03   
2                     ONEOK Inc.    787016807.0  7.074026e-03   1.414805e-03   
3          WEC Energy Group Inc.    732058286.0  6.580036e-03   1.316007e-03   
4                IHS Markit Ltd.    722329137.0  6.492587e-03   1.298517e-03   
..                           ...            ...           ...            ...   
243            Coty Inc. Class A     85716461.0  7.704543e-04   1.540909e-04   
244               AO Smith Corp.     82061091.0  7.375983e-04   1.475197e-04   
245         Lennar Corp. Class B      9520251.0  8.557187e-05   1.711437e-05   
246           News Corp. Class B      5167272.0  4.644553e-05   9.289107e-06   
247  Alliance Data Systems Corp.        23338.0  2.097714e-07   4.195428e-08   

     marketValue_y      weight_y  comb_weight_y  
0      711691839.0  1.331845e-03   2.663690e-04  
1      641115341.0  1.199769e-03   2.399539e-04  
2      624501862.0  1.168679e-03   2.337359e-04  
3      581278837.0  1.087793e-03   2.175585e-04  
4      603649797.0  1.129657e-03   2.259314e-04  
..             ...           ...            ...  
243     66573664.0  1.245845e-04   2.491690e-05  
244    130585480.0  2.443749e-04   4.887497e-05  
245      3440783.0  6.439007e-06   1.287801e-06  
246       213573.0  3.996759e-07   7.993518e-08  
247     91782405.0  1.717596e-04   3.435192e-05  

[248 rows x 7 columns]

In [82]:
df_vo_voo.holding.count() / dfvo.holding.count()

0.7315634218289085

In [83]:
df_vo_vti = dfvo.merge(dfvti, how='inner', left_on='holding', right_on='holding' )

In [84]:
df_vo_vti

holding  marketValue_x      weight_x  \
0               Newmont Goldcorp Corp.    896765462.0  8.060491e-03   
1               Amphenol Corp. Class A    807963791.0  7.262306e-03   
2                           ONEOK Inc.    787016807.0  7.074026e-03   
3                WEC Energy Group Inc.    732058286.0  6.580036e-03   
4                      IHS Markit Ltd.    722329137.0  6.492587e-03   
..                                 ...            ...           ...   
301               Datadog Inc. Class A     26213880.0  2.356210e-04   
302  Crowdstrike Holdings Inc. Class A     25919583.0  2.329757e-04   
303         Levi Strauss & Co. Class A     21514831.0  1.933840e-04   
304             ViacomCBS Inc. Class A      6668086.0  5.993546e-05   
305        Alliance Data Systems Corp.        23338.0  2.097714e-07   

     comb_weight_x  marketValue_y      weight_y  comb_weight_y  
0     1.612098e-03   1.004034e+09  1.121973e-03   2.243946e-04  
1     1.452461e-03   9.051708e+08  1.011496e-03   2.022993e-04  
2     1.414805e-03   8.799129e+08  9.832716e-04   1.966543e-04  
3     1.316007e-03   8.175716e+08  9.136075e-04   1.827215e-04  
4     1.298517e-03   8.076436e+08  9.025132e-04   1.805026e-04  
..             ...            ...           ...            ...  
301   4.712419e-05   2.940319e+07  3.285703e-05   6.571406e-06  
302   4.659514e-05   2.888251e+07  3.227519e-05   6.455037e-06  
303   3.867680e-05   2.445058e+07  2.732266e-05   5.464531e-06  
304   1.198709e-05   2.294200e+05  2.563688e-07   5.127375e-08  
305   4.195428e-08   1.380236e+08  1.542365e-04   3.084730e-05  

[306 rows x 7 columns]

In [85]:
df_vo_vti.holding.count() / dfvo.holding.count()

0.9026548672566371

In [86]:
df_vo_vti.sort_values(by=['weight_y'],ascending=False)

holding  marketValue_x  weight_x  \
12         Advanced Micro Devices Inc.    642678555.0  0.005777   
0               Newmont Goldcorp Corp.    896765462.0  0.008060   
1               Amphenol Corp. Class A    807963791.0  0.007262   
2                           ONEOK Inc.    787016807.0  0.007074   
110         AvalonBay Communities Inc.    368526570.0  0.003312   
..                                 ...            ...       ...   
301               Datadog Inc. Class A     26213880.0  0.000236   
302  Crowdstrike Holdings Inc. Class A     25919583.0  0.000233   
63             Liberty Broadband Corp.    458327941.0  0.004120   
303         Levi Strauss & Co. Class A     21514831.0  0.000193   
304             ViacomCBS Inc. Class A      6668086.0  0.000060   

     comb_weight_x  marketValue_y      weight_y  comb_weight_y  
12        0.001155   1.438734e+09  1.607734e-03   3.215468e-04  
0         0.001612   1.004034e+09  1.121973e-03   2.243946e-04  
1         0.001452   9.051708e+08  1.011496e-03   2.022993e-04  
2         0.001415   8.799129e+08  9.832716e-04   1.966543e-04  
110       0.000662   8.224050e+08  9.190086e-04   1.838017e-04  
..             ...            ...           ...            ...  
301       0.000047   2.940319e+07  3.285703e-05   6.571406e-06  
302       0.000047   2.888251e+07  3.227519e-05   6.455037e-06  
63        0.000824   2.849509e+07  3.184226e-05   6.368452e-06  
303       0.000039   2.445058e+07  2.732266e-05   5.464531e-06  
304       0.000012   2.294200e+05  2.563688e-07   5.127375e-08  

[306 rows x 7 columns]

In [87]:
df_vo_vti

holding  marketValue_x      weight_x  \
0               Newmont Goldcorp Corp.    896765462.0  8.060491e-03   
1               Amphenol Corp. Class A    807963791.0  7.262306e-03   
2                           ONEOK Inc.    787016807.0  7.074026e-03   
3                WEC Energy Group Inc.    732058286.0  6.580036e-03   
4                      IHS Markit Ltd.    722329137.0  6.492587e-03   
..                                 ...            ...           ...   
301               Datadog Inc. Class A     26213880.0  2.356210e-04   
302  Crowdstrike Holdings Inc. Class A     25919583.0  2.329757e-04   
303         Levi Strauss & Co. Class A     21514831.0  1.933840e-04   
304             ViacomCBS Inc. Class A      6668086.0  5.993546e-05   
305        Alliance Data Systems Corp.        23338.0  2.097714e-07   

     comb_weight_x  marketValue_y      weight_y  comb_weight_y  
0     1.612098e-03   1.004034e+09  1.121973e-03   2.243946e-04  
1     1.452461e-03   9.051708e+08  1.011496e-03   2.022993e-04  
2     1.414805e-03   8.799129e+08  9.832716e-04   1.966543e-04  
3     1.316007e-03   8.175716e+08  9.136075e-04   1.827215e-04  
4     1.298517e-03   8.076436e+08  9.025132e-04   1.805026e-04  
..             ...            ...           ...            ...  
301   4.712419e-05   2.940319e+07  3.285703e-05   6.571406e-06  
302   4.659514e-05   2.888251e+07  3.227519e-05   6.455037e-06  
303   3.867680e-05   2.445058e+07  2.732266e-05   5.464531e-06  
304   1.198709e-05   2.294200e+05  2.563688e-07   5.127375e-08  
305   4.195428e-08   1.380236e+08  1.542365e-04   3.084730e-05  

[306 rows x 7 columns]

In [88]:
df_vo_vti.groupby(by = 'weight_x').apply(lambda d: d.groupby('weight_y').size().nlargest())


weight_x      weight_y    
2.097714e-07  1.542365e-04    1
5.993546e-05  2.563688e-07    1
1.933840e-04  2.732266e-05    1
2.329757e-04  3.227519e-05    1
2.356210e-04  3.285703e-05    1
                             ..
6.492587e-03  9.025132e-04    1
6.580036e-03  9.136075e-04    1
7.074026e-03  9.832716e-04    1
7.262306e-03  1.011496e-03    1
8.060491e-03  1.121973e-03    1
Length: 306, dtype: int64

In [89]:
df_vo_vxf = dfvo.merge(dfvxf, how='inner', left_on='holding', right_on='holding' )

In [90]:
df_vo_vxf

holding  marketValue_x  weight_x  comb_weight_x  \
0            Lululemon Athletica Inc.    682273247.0  0.006133       0.001227   
1                         Splunk Inc.    581904975.0  0.005230       0.001046   
2                   CoStar Group Inc.    551726533.0  0.004959       0.000992   
3             Palo Alto Networks Inc.    541265600.0  0.004865       0.000973   
4                         DexCom Inc.    503501857.0  0.004526       0.000905   
..                                ...            ...       ...            ...   
86     Interactive Brokers Group Inc.     42699258.0  0.000384       0.000077   
87                         Chewy Inc.     31127237.0  0.000280       0.000056   
88               Datadog Inc. Class A     26213880.0  0.000236       0.000047   
89  Crowdstrike Holdings Inc. Class A     25919583.0  0.000233       0.000047   
90         Levi Strauss & Co. Class A     21514831.0  0.000193       0.000039   

    marketValue_y  weight_y  comb_weight_y  
0     393325388.0  0.005333       0.001067  
1     293984731.0  0.003986       0.000797  
2     310802491.0  0.004214       0.000843  
3     321314013.0  0.004357       0.000871  
4     283051310.0  0.003838       0.000768  
..            ...       ...            ...  
86     51246569.0  0.000695       0.000139  
87     22094259.0  0.000300       0.000060  
88     14743607.0  0.000200       0.000040  
89     14805605.0  0.000201       0.000040  
90     11768829.0  0.000160       0.000032  

[91 rows x 7 columns]

In [91]:
df_vo_vxf.holding.count() / dfvo.holding.count()

0.26843657817109146

## <b>Besides VO, both VTI and VOO containing high proportion of Mid-size companies - 90% and 73% respectively.</b>

## <b>Proporation of ETF also containing Small-size companies</b>

In [92]:
df_vb_vo = dfvb.merge(dfvo, how='inner', left_on='holding', right_on='holding' )

In [93]:
df_vb_vo

holding  marketValue_x  weight_x  comb_weight_x  \
0           Burlington Stores Inc.    188613190.0  0.001968       0.000394   
1         RingCentral Inc. Class A    157512817.0  0.001643       0.000329   
2     Jack Henry & Associates Inc.    142529937.0  0.001487       0.000297   
3                Vail Resorts Inc.    122721011.0  0.001280       0.000256   
4      Alliance Data Systems Corp.    122283863.0  0.001276       0.000255   
5               DXC Technology Co.    119153158.0  0.001243       0.000249   
6                  NRG Energy Inc.    118817321.0  0.001240       0.000248   
7          Jones Lang LaSalle Inc.    114267801.0  0.001192       0.000238   
8                     Cognex Corp.    112210181.0  0.001171       0.000234   
9                      Pentair plc     97582003.0  0.001018       0.000204   
10   Reliance Steel & Aluminum Co.     96732068.0  0.001009       0.000202   
11  People's United Financial Inc.     95936534.0  0.001001       0.000200   
12              XPO Logistics Inc.     91110729.0  0.000951       0.000190   
13  Robert Half International Inc.     88143696.0  0.000920       0.000184   
14           Cabot Oil & Gas Corp.     88123936.0  0.000919       0.000184   
15               FLIR Systems Inc.     86586786.0  0.000903       0.000181   
16     Axalta Coating Systems Ltd.     81578248.0  0.000851       0.000170   
17                  AO Smith Corp.     80515078.0  0.000840       0.000168   
18             Carvana Co. Class A     58232303.0  0.000608       0.000122   

    marketValue_y      weight_y  comb_weight_y  
0     190030853.0  1.708074e-03   3.416149e-04  
1     158720663.0  1.426646e-03   2.853291e-04  
2     141039879.0  1.267724e-03   2.535447e-04  
3     121630024.0  1.093260e-03   2.186520e-04  
4         23338.0  2.097714e-07   4.195428e-08  
5     120928383.0  1.086953e-03   2.173907e-04  
6     119544627.0  1.074516e-03   2.149031e-04  
7     112707607.0  1.013062e-03   2.026123e-04  
8     114316444.0  1.027522e-03   2.055045e-04  
9      96885880.0  8.708495e-04   1.741699e-04  
10     95332313.0  8.568854e-04   1.713771e-04  
11     94077095.0  8.456030e-04   1.691206e-04  
12     92421156.0  8.307188e-04   1.661438e-04  
13     87477338.0  7.862817e-04   1.572563e-04  
14     89336943.0  8.029966e-04   1.605993e-04  
15     87826573.0  7.894208e-04   1.578842e-04  
16     80769730.0  7.259910e-04   1.451982e-04  
17     82061091.0  7.375983e-04   1.475197e-04  
18     58196403.0  5.230928e-04   1.046186e-04

In [94]:
df_vb_vo.holding.count() / dfvb.holding.count()

0.01400147383935151

In [95]:
df_vb_voo = dfvb.merge(dfvoo, how='inner', left_on='holding', right_on='holding' )

In [96]:
df_vb_voo

holding  marketValue_x  weight_x  \
0                    Leidos Holdings Inc.    347521932.0  0.003626   
1                Zebra Technologies Corp.    344245760.0  0.003591   
2                      Atmos Energy Corp.    334201885.0  0.003487   
3                              IDEX Corp.    327868292.0  0.003421   
4                              STERIS plc    322979504.0  0.003370   
5                            Allegion plc    290041331.0  0.003026   
6                        PerkinElmer Inc.    269851193.0  0.002815   
7                           Assurant Inc.    193115445.0  0.002015   
8   Apartment Investment & Management Co.    192954483.0  0.002013   
9                    Leggett & Platt Inc.    166601221.0  0.001738   
10                        Flowserve Corp.    160909546.0  0.001679   
11                       Sealed Air Corp.    153617579.0  0.001603   
12          Wyndham Hotels & Resorts Inc.    148694447.0  0.001551   
13                   Quanta Services Inc.    144563083.0  0.001508   
14           Jack Henry & Associates Inc.    142529937.0  0.001487   
15                   Harley-Davidson Inc.    142482960.0  0.001486   
16                    Capri Holdings Ltd.    137187514.0  0.001431   
17                       Hanesbrands Inc.    134264775.0  0.001401   
18                             Unum Group    134071265.0  0.001399   
19                     Cimarex Energy Co.    131829685.0  0.001375   
20                            Macy's Inc.    130572971.0  0.001362   
21                               Gap Inc.    122617890.0  0.001279   
22            Alliance Data Systems Corp.    122283863.0  0.001276   
23                     DXC Technology Co.    119153158.0  0.001243   
24                        NRG Energy Inc.    118817321.0  0.001240   
25                         H&R Block Inc.    117557152.0  0.001226   
26                 Helmerich & Payne Inc.    115396425.0  0.001204   
27                         Nordstrom Inc.    110137268.0  0.001149   
28                          L Brands Inc.    105772691.0  0.001104   
29                            Pentair plc     97582003.0  0.001018   
30                           Macerich Co.     97126579.0  0.001013   
31         People's United Financial Inc.     95936534.0  0.001001   
32         Robert Half International Inc.     88143696.0  0.000920   
33                  Cabot Oil & Gas Corp.     88123936.0  0.000919   
34                      FLIR Systems Inc.     86586786.0  0.000903   
35                         AO Smith Corp.     80515078.0  0.000840   

    comb_weight_x  marketValue_y      weight_y  comb_weight_y  
0        0.000725    260170084.0  4.868767e-04   9.737534e-05  
1        0.000718    266652028.0  4.990069e-04   9.980137e-05  
2        0.000697    266629160.0  4.989641e-04   9.979281e-05  
3        0.000684    261269204.0  4.889336e-04   9.778671e-05  
4        0.000674    258211674.0  4.832118e-04   9.664235e-05  
5        0.000605    231312875.0  4.328739e-04   8.657478e-05  
6        0.000563    215386152.0  4.030690e-04   8.061380e-05  
7        0.000403    151696131.0  2.838809e-04   5.677617e-05  
8        0.000403    153747588.0  2.877199e-04   5.754398e-05  
9        0.000348    133603402.0  2.500225e-04   5.000450e-05  
10       0.000336    130032437.0  2.433399e-04   4.866798e-05  
11       0.000321    122863362.0  2.299238e-04   4.598477e-05  
12       0.000310        13567.0  2.538899e-08   5.077798e-09  
13       0.000302    115721961.0  2.165596e-04   4.331192e-05  
14       0.000297    223846719.0  4.189019e-04   8.378038e-05  
15       0.000297    114756848.0  2.147535e-04   4.295070e-05  
16       0.000286    115439611.0  2.160312e-04   4.320624e-05  
17       0.000280    107179741.0  2.005739e-04   4.011477e-05  
18       0.000280    120278731.0  2.250870e-04   4.501740e-05  
19       0.000275    106862869.0  1.999809e-04   3.999617e-05  
20       0.000272    104839714.0  1.961948e-04   3.923896e-05  
21       0.000256     75081621.0  1

In [97]:
df_vb_voo.holding.count() / dfvb.holding.count()

0.026529108327192335

In [98]:
df_vb_vti = dfvb.merge(dfvti, how='inner', left_on='holding', right_on='holding' )

In [99]:
df_vb_vti

holding  marketValue_x      weight_x  \
0                  Leidos Holdings Inc.    347521932.0  3.625613e-03   
1              Zebra Technologies Corp.    344245760.0  3.591434e-03   
2                    Atmos Energy Corp.    334201885.0  3.486648e-03   
3                            IDEX Corp.    327868292.0  3.420571e-03   
4                            STERIS plc    322979504.0  3.369568e-03   
...                                 ...            ...           ...   
1231          AMAG Pharmaceuticals Inc.        33139.0  3.457313e-07   
1232  Altisource Portfolio Solutions SA        19775.0  2.063078e-07   
1233       Synergy Pharmaceuticals Inc.        12807.0  1.336124e-07   
1234                       Ashford Inc.          614.0  6.405715e-09   
1235                    Dorian LPG Ltd.          310.0  3.234156e-09   

      comb_weight_x  marketValue_y      weight_y  comb_weight_y  
0      7.251226e-04    391009662.0  4.369395e-04   8.738790e-05  
1      7.182867e-04    388356671.0  4.339749e-04   8.679498e-05  
2      6.973297e-04    373936347.0  4.178607e-04   8.357213e-05  
3      6.841143e-04    367400256.0  4.105568e-04   8.211136e-05  
4      6.739136e-04    363860225.0  4.066010e-04   8.132019e-05  
...             ...            ...           ...            ...  
1231   6.914625e-08     10748629.0  1.201121e-05   2.402243e-06  
1232   4.126157e-08      2110971.0  2.358936e-06   4.717871e-07  
1233   2.672247e-08        32948.0  3.681823e-08   7.363645e-09  
1234   1.281143e-09      1001985.0  1.119683e-06   2.239366e-07  
1235   6.468312e-10          310.0  3.464141e-10   6.928281e-11  

[1236 rows x 7 columns]

In [100]:
df_vb_vti.holding.count() / dfvb.holding.count()

0.9108327192336035

In [101]:
df_vb_vti.sort_values(by=['weight_y'],ascending=False)

holding  marketValue_x      weight_x  \
63          Burlington Stores Inc.    188613190.0  1.967756e-03   
0             Leidos Holdings Inc.    347521932.0  3.625613e-03   
1         Zebra Technologies Corp.    344245760.0  3.591434e-03   
2               Atmos Energy Corp.    334201885.0  3.486648e-03   
3                       IDEX Corp.    327868292.0  3.420571e-03   
...                            ...            ...           ...   
1229        FTS International Inc.        77794.0  8.116062e-07   
1230                    Unit Corp.        47427.0  4.947945e-07   
1225        Vir Biotechnology Inc.       549100.0  5.728629e-06   
1233  Synergy Pharmaceuticals Inc.        12807.0  1.336124e-07   
1235               Dorian LPG Ltd.          310.0  3.234156e-09   

      comb_weight_x  marketValue_y      weight_y  comb_weight_y  
63     3.935513e-04    425579686.0  4.755703e-04   9.511406e-05  
0      7.251226e-04    391009662.0  4.369395e-04   8.738790e-05  
1      7.182867e-04    388356671.0  4.339749e-04   8.679498e-05  
2      6.973297e-04    373936347.0  4.178607e-04   8.357213e-05  
3      6.841143e-04    367400256.0  4.105568e-04   8.211136e-05  
...             ...            ...           ...            ...  
1229   1.623212e-07       904394.0  1.010628e-06   2.021257e-07  
1230   9.895891e-08       890950.0  9.956052e-07   1.991210e-07  
1225   1.145726e-06       595753.0  6.657330e-07   1.331466e-07  
1233   2.672247e-08        32948.0  3.681823e-08   7.363645e-09  
1235   6.468312e-10          310.0  3.464141e-10   6.928281e-11  

[1236 rows x 7 columns]

In [102]:
df_vb_vxf = dfvb.merge(dfvxf, how='inner', left_on='holding', right_on='holding' )

In [103]:
df_vb_vxf

holding  marketValue_x      weight_x  \
0              Zebra Technologies Corp.    344245760.0  3.591434e-03   
1            Teledyne Technologies Inc.    317065733.0  3.307871e-03   
2      Equity LifeStyle Properties Inc.    305528499.0  3.187506e-03   
3               Tyler Technologies Inc.    293532068.0  3.062350e-03   
4                         Teradyne Inc.    286874171.0  2.992890e-03   
...                                 ...            ...           ...   
1315                         Unit Corp.        47427.0  4.947945e-07   
1316          AMAG Pharmaceuticals Inc.        33139.0  3.457313e-07   
1317  Altisource Portfolio Solutions SA        19775.0  2.063078e-07   
1318                       Ashford Inc.          614.0  6.405715e-09   
1319                    Dorian LPG Ltd.          310.0  3.234156e-09   

      comb_weight_x  marketValue_y      weight_y  comb_weight_y  
0      7.182867e-04        33718.0  4.571922e-07   9.143844e-08  
1      6.615742e-04    180019213.0  2.440933e-03   4.881866e-04  
2      6.375011e-04    182628826.0  2.476318e-03   4.952635e-04  
3      6.124700e-04    165806953.0  2.248225e-03   4.496450e-04  
4      5.985779e-04    161877127.0  2.194939e-03   4.389878e-04  
...             ...            ...           ...            ...  
1315   9.895891e-08       574155.0  7.785135e-06   1.557027e-06  
1316   6.914625e-08      6189674.0  8.392760e-05   1.678552e-05  
1317   4.126157e-08      2953121.0  4.004223e-05   8.008446e-06  
1318   1.281143e-09       462536.0  6.271661e-06   1.254332e-06  
1319   6.468312e-10      6788893.0  9.205258e-05   1.841052e-05  

[1320 rows x 7 columns]

In [104]:
df_vb_vxf.holding.count() / dfvb.holding.count()

0.9727339719970524

## <b>Besides VO, both VTI and VOO containing high proportion of Mid-size companies - 90% and 73% respectively.</b>

## <b>Besides VB, both VXF and VTI containing high proportion of Small-size companies - 97% and 91% respectively.</b>

In [105]:
df_vo_voo_vb = df_vo_voo.merge(dfvb, how='inner', left_on='holding', right_on='holding' )

In [106]:
df_vo_voo_vb

holding  marketValue_x      weight_x  comb_weight_x  \
0    Jack Henry & Associates Inc.    141039879.0  1.267724e-03   2.535447e-04   
1              DXC Technology Co.    120928383.0  1.086953e-03   2.173907e-04   
2                 NRG Energy Inc.    119544627.0  1.074516e-03   2.149031e-04   
3                     Pentair plc     96885880.0  8.708495e-04   1.741699e-04   
4  People's United Financial Inc.     94077095.0  8.456030e-04   1.691206e-04   
5           Cabot Oil & Gas Corp.     89336943.0  8.029966e-04   1.605993e-04   
6               FLIR Systems Inc.     87826573.0  7.894208e-04   1.578842e-04   
7  Robert Half International Inc.     87477338.0  7.862817e-04   1.572563e-04   
8                  AO Smith Corp.     82061091.0  7.375983e-04   1.475197e-04   
9     Alliance Data Systems Corp.        23338.0  2.097714e-07   4.195428e-08   

   marketValue_y  weight_y  comb_weight_y  marketValue    weight  comb_weight  
0    223846719.0  0.000419       0.000084  142529937.0  0.001487     0.000297  
1    192412798.0  0.000360       0.000072  119153158.0  0.001243     0.000249  
2    199939121.0  0.000374       0.000075  118817321.0  0.001240     0.000248  
3    154164254.0  0.000288       0.000058   97582003.0  0.001018     0.000204  
4    150079081.0  0.000281       0.000056   95936534.0  0.001001     0.000200  
5    142000922.0  0.000266       0.000053   88123936.0  0.000919     0.000184  
6    139647783.0  0.000261       0.000052   86586786.0  0.000903     0.000181  
7    148453209.0  0.000278       0.000056   88143696.0  0.000920     0.000184  
8    130585480.0  0.000244       0.000049   80515078.0  0.000840     0.000168  
9     91782405.0  0.000172       0.000034  122283863.0  0.001276     0.000255

In [107]:
df_vo_vti_vb = df_vo_vti.merge(dfvb, how='inner', left_on='holding', right_on='holding' )

In [108]:
df_vo_vti_vb

holding  marketValue_x      weight_x  \
0           Burlington Stores Inc.    190030853.0  1.708074e-03   
1     Jack Henry & Associates Inc.    141039879.0  1.267724e-03   
2                Vail Resorts Inc.    121630024.0  1.093260e-03   
3               DXC Technology Co.    120928383.0  1.086953e-03   
4                  NRG Energy Inc.    119544627.0  1.074516e-03   
5                     Cognex Corp.    114316444.0  1.027522e-03   
6          Jones Lang LaSalle Inc.    112707607.0  1.013062e-03   
7                      Pentair plc     96885880.0  8.708495e-04   
8    Reliance Steel & Aluminum Co.     95332313.0  8.568854e-04   
9   People's United Financial Inc.     94077095.0  8.456030e-04   
10              XPO Logistics Inc.     92421156.0  8.307188e-04   
11           Cabot Oil & Gas Corp.     89336943.0  8.029966e-04   
12               FLIR Systems Inc.     87826573.0  7.894208e-04   
13  Robert Half International Inc.     87477338.0  7.862817e-04   
14                  AO Smith Corp.     82061091.0  7.375983e-04   
15     Axalta Coating Systems Ltd.     80769730.0  7.259910e-04   
16     Alliance Data Systems Corp.        23338.0  2.097714e-07   

    comb_weight_x  marketValue_y  weight_y  comb_weight_y  marketValue  \
0    3.416149e-04    425579686.0  0.000476       0.000095  188613190.0   
1    2.535447e-04    316508571.0  0.000354       0.000071  142529937.0   
2    2.186520e-04    271452305.0  0.000303       0.000061  122721011.0   
3    2.173907e-04    271710143.0  0.000304       0.000061  119153158.0   
4    2.149031e-04    268842606.0  0.000300       0.000060  118817321.0   
5    2.055045e-04    256145839.0  0.000286       0.000057  112210181.0   
6    2.026123e-04    251237809.0  0.000281       0.000056  114267801.0   
7    1.741699e-04    217972818.0  0.000244       0.000049   97582003.0   
8    1.713771e-04    212900226.0  0.000238       0.000048   96732068.0   
9    1.691206e-04    212008675.0  0.000237       0.000047   95936534.0   
10   1.661438e-04    206690234.0  0.000231       0.000046   91110729.0   
11   1.605993e-04    200358058.0  0.000224       0.000045   88123936.0   
12   1.578842e-04    197052719.0  0.000220       0.000044   86586786.0   
13   1.572563e-04    196328614.0  0.000219       0.000044   88143696.0   
14   1.475197e-04    184140605.0  0.000206       0.000041   80515078.0   
15   1.451982e-04    179425056.0  0.000201       0.000040   81578248.0   
16   4.195428e-08    138023615.0  0.000154       0.000031  122283863.0   

      weight  comb_weight  
0   0.001968     0.000394  
1   0.001487     0.000297  
2   0.001280     0.000256  
3   0.001243     0.000249  
4   0.001240     0.000248  
5   0.001171     0.000234  
6   0.001192     0.000238  
7   0.001018     0.000204  
8   0.001009     0.000202  
9   0.001001     0.000200  
10  0.000951     0.000190  
11  0.000919     0.000184  
12  0.000903     0.000181  
13  0.000920     0.000184  
14  0.000840     0.000168  
15  0.000851     0.000170  
16  0.001276     0.000255

In [109]:
df_vb_vxf_vo = df_vb_vxf.merge(dfvo, how='inner', left_on='holding', right_on='holding' )

In [110]:
df_vb_vxf_vo

holding  marketValue_x  weight_x  comb_weight_x  \
0         Burlington Stores Inc.    188613190.0  0.001968       0.000394   
1       RingCentral Inc. Class A    157512817.0  0.001643       0.000329   
2              Vail Resorts Inc.    122721011.0  0.001280       0.000256   
3        Jones Lang LaSalle Inc.    114267801.0  0.001192       0.000238   
4                   Cognex Corp.    112210181.0  0.001171       0.000234   
5  Reliance Steel & Aluminum Co.     96732068.0  0.001009       0.000202   
6             XPO Logistics Inc.     91110729.0  0.000951       0.000190   
7    Axalta Coating Systems Ltd.     81578248.0  0.000851       0.000170   
8            Carvana Co. Class A     58232303.0  0.000608       0.000122   

   marketValue_y  weight_y  comb_weight_y  marketValue    weight  comb_weight  
0    214106488.0  0.002903       0.000581  190030853.0  0.001708     0.000342  
1    179453748.0  0.002433       0.000487  158720663.0  0.001427     0.000285  
2    138128410.0  0.001873       0.000375  121630024.0  0.001093     0.000219  
3    128130588.0  0.001737       0.000347  112707607.0  0.001013     0.000203  
4    135976297.0  0.001844       0.000369  114316444.0  0.001028     0.000206  
5    113415714.0  0.001538       0.000308   95332313.0  0.000857     0.000171  
6    104224886.0  0.001413       0.000283   92421156.0  0.000831     0.000166  
7     89794973.0  0.001218       0.000244   80769730.0  0.000726     0.000145  
8     65355960.0  0.000886       0.000177   58196403.0  0.000523     0.000105

## <b>There does not seem to be a ETF containing both Mid-sized and Small-sized companies which makes sense. </b>

## <b>Besides VO, both VTI and VOO containing high proportion of Mid-size companies - 90% and 73% respectively.</b>

## <b>Besides VO, both VTI and VOO containing high proportion of Mid-size companies - 90% and 73% respectively.</b>

## <b>Based on comparing the inner join results between ETFs with mid-size and small-size ETC, to get broader exposure to more mid and small size companies, there will likely include VTI which matches 90% of mid-size ETC and 91% of small-size ETC. Though VXF has high exposure (97%) of small-size ETC, there is just matches 26% of mid-size ETC. The mid-size VO and small-size VB have about 14% of common companies. So investing in either one will lower exposure of the other. </b>
## <b>As a result, without concerning the weight, there will divide most of investment into VTI, about equal portion in VXF, VB and VO. The portion can be 70% of VTI, and 10% of each of XVF, VB and VO.</b>